src/kernels/types.ts

```diff
@@ -94,7 +95,12 @@ export type PythonKernelConnectionMetadata = Readonly<{
  * This ensure we don't update is somewhere unnecessarily (such updates would be unexpected).
  * Unexpected as connections are defined once & not changed, if we need to change then user needs to create a new connection.
  */
-export type KernelConnectionMetadata = RemoteKernelConnectionMetadata | LocalKernelConnectionMetadata;
+export type KernelConnectionMetadata = (RemoteKernelConnectionMetadata | LocalKernelConnectionMetadata) & {
+    /**
+     * Kernel Connections may optionally be tagged with the info for the kernel finder that contributed them
+     */
+    kernelFinderInfo?: IContributedKernelFinderInfo;
```

Comment:
Why not make them mandatory? Based on the code this looks like they will always be present (execpt for the cases where we load these from cache, and we can still update them from within the finders)

Reply:
I'll take a second look here. Initially the reason that I did this was because we manually register controllers from outside the main finder path at a few places where they don't come from the finders. I don't love that we do this as it complicates things, so maybe I can get rid of these cases or manually put in some type of fake KernelFinderInfo. If those paths seem like they would mess with the stable path functionality I'll leave this as optional.

Reply:
i still think it could cause issues
if you check, these types are all read-only.
we want to prevent updating properties during runtime.
optionally you could have another class/dictionary to map the info by the kernel connection id

ie i don't think so should add these write properties, we had a lot of issues in the party as we didn't know when it could change

Reply:
If you are able to make this a readonly property can you add this to the individual types instead of adding this to the top level type.
I.e. add the property to `LocalKernelSpecConnectionMetadata` itself (add to the 4 different types).

After all, if you look at an individual connection that should contain the `kernelFinderInfo`, but based on the above types it will not, as the proeprty is added to the top level union.